### ISSUE TITLE
EventCounters not reporting data in 2.0

### ISSUE THREAD

**Issue by anurse** [dotnet/corefx#24273]
I have a sample app that uses EventCounters: https://github.com/anurse/EventCounterTest

In the app, I write to an event counter every 100ms, and I configure a listener to aggregate event counters every 1s:

```csharp
using System;
using System.Collections.Generic;
using System.Diagnostics.Tracing;

namespace EventCounterTest
{
    class Program
    {
        static void Main(string[] args) => AsyncMain(args).Wait();

        static async Task AsyncMain(string[] args)
        {
            var cts = new CancellationTokenSource();
            Console.CancelKeyPress += (sender, a) =>
            {
                cts.Cancel();
                a.Cancel = true;
            };

            var listener = new HostingEventSourceListener();

            Console.WriteLine("Press Ctrl-C to stop");
            for (var i = 0; !cts.Token.IsCancellationRequested; i++)
            {
                var url = $"https://localhost/?iter={i}";

                MinimalEventCounterSource.Log.Request(url, i);
                try
                {
                    await Task.Delay(100, cts.Token);
                }
                catch (OperationCanceledException)
                {
                    return;
                }
            }
        }
    }

    [EventSource(Name = "Samples-EventCounterDemos-Minimal")]
    public sealed class MinimalEventCounterSource : EventSource
    {
        public static MinimalEventCounterSource Log = new MinimalEventCounterSource();
        private EventCounter _requestCounter;

        private MinimalEventCounterSource()
        {
            _requestCounter = new EventCounter("request", this);
        }

        /// <summary>
        /// Call this method to indicate that a request for a URL was made which tool a particular amount of time
        /// </summary>
        public void Request(string url, float elapsedMSec)
        {
            WriteEvent(1, url, elapsedMSec);    // This logs it to the event stream if events are on.    
            _requestCounter.WriteMetric(elapsedMSec);        // This adds it to the EventCounter called 'Request' if PerfCounters are on
        }
    }

    public class HostingEventSourceListener : EventListener
    {
        protected override void OnEventSourceCreated(EventSource eventSource)
        {
            if (eventSource.Name.Equals("Samples-EventCounterDemos-Minimal"))
            {
                EnableEvents(eventSource, EventLevel.LogAlways, EventKeywords.All, new Dictionary<string, string>()
                {
                    { "EventCounterIntervalSec", "1" }
                });
            }
        }

        protected override void OnEventWritten(EventWrittenEventArgs eventData)
        {
            Console.WriteLine($"{eventData.EventName}");
            foreach (var payload in eventData.Payload)
            {
                if (payload is IDictionary<string, object> payloadDictionary)
                {
                    foreach (var data in payloadDictionary)
                    {
                        Console.WriteLine($" {data.Key} - {data.Value}");
                    }
                }
            }
        }
    }
}
```

When I run the app in `netcoreapp1.1`, I get the following output:"

```
Press Ctrl-C to stop
Request
Request
Request
Request
Request
Request
Request
Request
Request
Request
EventCounters
 Name - request
 Mean - 4.5
 StandardDerivation - 2.872281
 Count - 10
 Min - 0
 Max - 9
 IntervalSec - 1.015035
Request
...
```

So the EventCounter is firing at the requested interval.

However, if I switch the target framework to `netcoreapp2.0` and re-run the app, I get this output:

```
Press Ctrl-C to stop
Request
Request
Request
Request
Request
Request
Request
Request
Request
Request
Request
...
```

The EventCounters event is never fired. The same behavior is seen with out-of-process monitoring in PerfView. The only difference between the projects is the target framework.

I investigated a little further and noticed that [in 2.0.0](https://github.com/dotnet/corefx/blob/release/2.0.0/src/System.Diagnostics.Tracing/src/System/Diagnostics/Tracing/EventCounter.cs#L246) the code that attaches the EventCounter to the EventSource is `#ifdef`ed:

https://github.com/dotnet/corefx/blob/653cb2fc7844dddeff43b03638bc0140bc4fc61c/src/System.Diagnostics.Tracing/src/System/Diagnostics/Tracing/EventCounter.cs#L248

However, when I view the assembly in my 2.0.0 shared runtime folder in ILspy, I see that the method is empty (Looking at "C:\Program Files\dotnet\shared\Microsoft.NETCore.App\2.0.0\System.Diagnostics.Tracing.dll", and I verified against the DLL in the ZIP on https://dot.net as well):

![image](https://user-images.githubusercontent.com/7574/30891472-71614b18-a2e8-11e7-944e-ad74cd5bce02.png)

Looking at 1.1.x, I see the code is present (Looking at "C:\Program Files\dotnet\shared\Microsoft.NETCore.App\1.1.4\System.Diagnostics.Tracing.dll"):

![image](https://user-images.githubusercontent.com/7574/30891490-8c0bf346-a2e8-11e7-8c09-6bfb7250ceee.png)

It looks like maybe this code was `#ifdef`ed out of the 2.0 build :(

/cc @DamianEdwards @davidfowl @JunTaoLuo @vancem

**Comment by davidfowl** [comment]
bump @brianrob

**Comment by vancem** [comment]
There is a reasonably subtle versioning issue associated with the EventCommandExecuted callback (Basically not all versions of .NET have it).   I will see if I can figure out what is going on later today.

**Comment by anurse** [comment]
Yeah, I noticed that code when looking at the `#ifdef` and at the more recent version of the code. But it appears that for some reason the define was never set when doing the build that ended up in .NET Core 2.0.

**Comment by vancem** [comment]
I have looked into this a bit.   

I actually don't know why the code in the release branch is wrong.  The #ifdef *should* be active but it is not.    In fixing other issues, I changed this to be simpler/more maintainable, and that DOES work.   

Anyway, the change that fixes this in the current branch is not in the release branch.   It seems to be a classic case of a fix being fixed in the master branch but not being integrated to the release branch, so it was missing in the release.    This is the commit that is missing that would have fixed this. 

18be0c74	 7/24/2017 9:44:50 PM Fixed races in EventCounters.  Preped to add Dispose() method

Thus merging this into the release branch should fix the issue.    I will check with people owning the release bar about getting this in.
